lib/send.js

```diff
@@ -0,0 +1,685 @@
+  let offset = options.start ?? 0
+
+  let statusCode = 200
+  const headers = {}
+
+  debug('send "%s"', path)
+
+  // set header fields
+  if (options.acceptRanges) {
+    debug('accept ranges')
+    headers['Accept-Ranges'] = 'bytes'
+  }
+
+  if (options.cacheControl) {
+    let cacheControl = 'public, max-age=' + Math.floor(options.maxage / 1000)
+
+    if (options.immutable) {
+      cacheControl += ', immutable'
+    }
+
+    debug('cache-control %s', cacheControl)
+    headers['Cache-Control'] = cacheControl
+  }
+
+  if (options.lastModified) {
+    const modified = stat.mtime.toUTCString()
+    debug('modified %s', modified)
+    headers['Last-Modified'] = modified
+  }
+
+  if (options.etag) {
+    const etag = 'W/"' + stat.size.toString(16) + '-' + stat.mtime.getTime().toString(16) + '"'
+    debug('etag %s', etag)
+    headers.ETag = etag
+  }
+
+  // set content-type
+  let type = mime.getType(path) || mime.default_type
+  debug('content-type %s', type)
+  if (type && isUtf8MimeType(type)) {
+    type += '; charset=UTF-8'
+  }
+  if (type) {
+    headers['Content-Type'] = type
+  }
+
+  // conditional GET support
+  if (isConditionalGET(request)) {
+    if (isPreconditionFailure(request, headers)) {
+      return sendError(412)
+    }
+
+    if (isNotModifiedFailure(request, headers)) {
+      return sendNotModified(headers)
+    }
+  }
+
+  // adjust len to start/end options
+  len = Math.max(0, len - offset)
+  if (options.end !== undefined) {
+    const bytes = options.end - offset + 1
+    if (len > bytes) len = bytes
+  }
+
+  // Range support
+  if (options.acceptRanges) {
+    const rangeHeader = request.headers.range
+
+    if (
+      rangeHeader !== undefined &&
+      BYTES_RANGE_REGEXP.test(rangeHeader)
+    ) {
+      // If-Range support
+      if (isRangeFresh(request, headers)) {
+        // parse
+        const ranges = parseBytesRange(len, rangeHeader)
+
+        // unsatisfiable
+        if (ranges.length === 0) {
+          debug('range unsatisfiable')
+
+          // Content-Range
+          headers['Content-Range'] = contentRange('bytes', len)
+
+          // 416 Requested Range Not Satisfiable
+          return sendError(416, {
+            headers: { 'Content-Range': headers['Content-Range'] }
+          })
+          // valid (syntactically invalid/multiple ranges are treated as a regular response)
+        } else if (ranges.length === 1) {
+          debug('range %j', ranges)
+
+          // Content-Range
+          statusCode = 206
+          headers['Content-Range'] = contentRange('bytes', len, ranges[0])
+
+          // adjust for requested range
+          offset += ranges[0].start
+          len = ranges[0].end - ranges[0].start + 1
+        }
+      } else {
+        debug('range stale')
+      }
+    }
+  }
+
+  // content-length
+  headers['Content-Length'] = len
+
+  // HEAD support
+  if (request.method === 'HEAD') {
+    return { statusCode, headers, stream: Readable.from('') }
```

Comment:
Not a huge issue, but is there a reason for the empty `Readable` here and not `null`?

Maybe to keep the usage simple (`stream` always pipeable), but it forces  the user to always check the statusCode if they don't want to pipe in [these](https://github.com/fastify/send/blob/bc0be1fa9ff8cf044619b1450c1541917d6bd773/lib/send.js#L442) cases, which is fine, but I think if we just returned `null`, a generic `if` check on `stream` would be faster in higher level libraries like `@fastify/static`

```suggestion
    stream: null
```

Also closer to [Matteo's proposal](https://github.com/fastify/send/issues/15#issuecomment-1386877019):

```js
import http from 'http'
import send from '@fastify/send'
import { pipeline } from 'stream'

const app = http.createServer(async function (req, res) {
  // the options include the duplex options
  const { statusCode, headers, stream } = await send(req, req.url, { root: './' })
  res.writeHead(statusCode, headers)
  if (stream) {
    pipeline(stream, res, function(err) {
      console.log(err)
    })
  } else {
    res.end()
  }
})
```

Reply:
I would like to provide a consistent output.
It should always provide `statusCode`, `headers` and `stream`. 
So the users doesn't need to care about when and why the response isn't there.